ibis/pyspark/compiler.py

```diff
@@ -1476,7 +1477,7 @@ def compile_not_null(t, expr, scope, **kwargs):
 @compiles(ops.ElementWiseVectorizedUDF)
 def compile_elementwise_udf(t, expr, scope):
     op = expr.op()
-    spark_output_type = ibis_dtype_to_spark_dtype(op._output_type)
+    spark_output_type = spark_dtype(op._output_type)
```

Comment:
Why this change? What's the difference between `ibis_dtype_to_spark_dtype` and `spark_dtype`? It is confusing

Reply:
Basically, `ibis_dtype_to_spark_dtype` doesn't handle all Ibis DataTypes

`spark_dtype` is the "main"/generic function that converts an Ibis type to a Spark type

`ibis_dtype_to_spark_dtype` is one of the many functions that overloads it—This one's used specifically when the argument is an Ibis DataType in [this list](https://github.com/ibis-project/ibis/blob/f92adea724da4c6a1af3887a8e3e0ff407e497a7/ibis/spark/datatypes.py#L10), but won't handle anything else

---
_More info (getting into the details):_

**["this list"](https://github.com/ibis-project/ibis/blob/f92adea724da4c6a1af3887a8e3e0ff407e497a7/ibis/spark/datatypes.py#L10)** = All the Ibis DataTypes and Spark types that can be "trivially" converted in either direction

There are some Ibis DataType **subclasses** that are not in the list. E.g. Ibis Decimal, because converting a Spark DecimalType to an Ibis Decimal is not "trivial" (can't just `return dt.Decimal()` and call it a day—we need to construct an Ibis Decimal with the correct `precision` and `scale` from the original Spark DecimalType). Essentially, `ibis_dtype_to_spark_dtype` covers **most** Ibis DataType subclasses but has exceptions. But using `spark_dtype` would cover those exceptions

Reply:
It's a little confusing, I'll see if I can come up with function names that are clearer. It is tricky though

Also I just realized Timestamp I think actually belongs in the list of types that `ibis_dtype_to_spark_dtype` handles because both ways the conversion is trivial. I'll refactor that

Reply:
I won't make any changes to the function names in this PR because I'd have to slightly refactor code that's unrelated to this bugfix if I do

I think the best action though (for a separate PR) is to rename `ibis_dtype_to_spark_dtype` to `_ibis_dtype_to_spark_dtype` (basically, shouldn't be used outside of that module because it isn't clear exactly what types it can handle), and replace usages of `ibis_dtype_to_spark_dtype` with `spark_dtype`